presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestMergeIdenticalWindows.java

```diff
@@ -0,0 +1,150 @@
+                "sum(discount) over (partition by suppkey order by orderkey rows between unbounded preceding and current row) sum_discount_A " +
+                "from lineitem";
+
+        assertUnitPlan(sql,
+                anyTree(
+                        window(ImmutableList.of(functionCall("sum", "quantity"), functionCall("sum", "discount")),
+                                window(ImmutableList.of(functionCall("sum", "quantity")),
+                                        any()))));
+    }
+
+    @Test
+    public void testIdenticalWindowSpecificationsABcpA()
+    {
+        @Language("SQL") String sql = "select " +
+                "sum(quantity) over (partition by suppkey order by orderkey rows between unbounded preceding and current row) sum_quantity_A, " +
+                "lag(quantity, 1, 0.0) over (partition by orderkey order by shipdate rows between UNBOUNDED preceding and CURRENT ROW) sum_quantity_B, " +
+                "sum(discount) over (partition by suppkey order by orderkey rows between unbounded preceding and current row) sum_discount_A " +
+                "from lineitem";
+
+        assertUnitPlan(sql,
+                anyTree(
+                        window(ImmutableList.of(functionCall("sum", "discount")),
+                                window(ImmutableList.of(functionCall("lag", "quantity", "*", "*")),
+                                        project(
+                                                window(ImmutableList.of(functionCall("sum", "quantity")),
+                                                        any()))))));
+    }
+
+    @Test
+    public void testIdenticalWindowSpecificationsAAcpA()
+    {
+        @Language("SQL") String sql = "select " +
+                "sum(quantity) over (partition by suppkey order by orderkey rows between unbounded preceding and current row) sum_quantity_A, " +
+                "lag(quantity, 1, 0.0) over (partition by suppkey order by orderkey rows between UNBOUNDED preceding and CURRENT ROW) sum_quantity_B, " +
+                "sum(discount) over (partition by suppkey order by orderkey rows between unbounded preceding and current row) sum_discount_A " +
+                "from lineitem";
+
+        assertUnitPlan(sql,
+                anyTree(
+                        window(ImmutableList.of(functionCall("sum", "discount"), functionCall("lag", "quantity", "*", "*")),
+                                project(
+                                        window(ImmutableList.of(functionCall("sum", "quantity")),
+                                                any())))));
+    }
+
+    private void assertUnitPlan(@Language("SQL") String sql, PlanMatchPattern pattern)
+    {
+        Plan actualPlan = unitPlan(sql);
+        queryRunner.inTransaction(transactionSession -> {
+            PlanAssert.assertPlan(transactionSession, queryRunner.getMetadata(), actualPlan, pattern);
+            return null;
+        });
+    }
+
+    private Plan unitPlan(@Language("SQL") String sql)
```

Comment:
Oof. This was the subject of a lot of discussion at the Boston office with @maciejgrzybek and @cawallin.

The short version is that if you run all the optimizers you aren't really running a true unit test, and if you get a failure, you don't necessarily know whether it's because MIW is broken or because some other optimizer changed the plan.

With the two types of tests in here we can separately answer the questions of:

1. Is MIW working as it's supposed to be? The tests running the minimal set of optimizers can tell us this.
2. Has some other optimizer changed the plan? The test running the full set of optimizers answers this, though it isn't actually meaningful unless we know the answer to question 1 is "yes".

The tests that use the minimal set of optimizers are closer to true "unit" tests in that they verify the behavior of MIW with as few external dependencies as possible. Unfortunately, those dependencies include the parser and analyzer, hence "unit" in quotes.

The conclusion of the discussion here was that nobody really likes the fact that this isn't a true unit test, but that we aren't going to try to generalize a solution from a sample size of one.